### PR TITLE
AMDGPU: Don't avoid clamp of bit shift in BFE pattern

### PR DESCRIPTION
**Pull request by changpeng**
Enable pattern matching from "x<<32-y>>32-y" to "bfe x, 0, y" when we know y is in [0,31].
  This is the follow-up for the PR: https://github.com/llvm/llvm-project/pull/114279 to fix the issue: https://github.com/llvm/llvm-project/issues/114282

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIInstructions.td [hunk 1]
```diff
@@ -3553,6 +3553,23 @@ def : AMDGPUPat <
   (V_BFE_U32_e64 $src, (i32 0), $width)
 >;
 
+def uint5Bits : PatLeaf<(i32 VGPR_32:$width), [{
+  return CurDAG->computeKnownBits(SDValue(N, 0)).countMinLeadingZeros() >= 27;
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SIInstructions.td:3557]: `countMaxTrailingOnes() <= 5` would be easier to read.

**Review by rampitec — APPROVED** [review]
LGTM with a nit.

File: llvm/test/CodeGen/AMDGPU/extract-lowbits.ll [hunk 2]
```diff
@@ -150,22 +150,14 @@ define i32 @bzhi32_c4_commutative(i32 %val, i32 %numlowbits) nounwind {
...
-;
-; VI-LABEL: bzhi32_d0:
-; VI:       ; %bb.0:
-; VI-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
-; VI-NEXT:    v_sub_u32_e32 v1, vcc, 32, v1
-; VI-NEXT:    v_lshlrev_b32_e32 v0, v1, v0
-; VI-NEXT:    v_lshrrev_b32_e32 v0, v1, v0
-; VI-NEXT:    s_setpc_b64 s[30:31]
-  %numhighbits = sub i32 32, %numlowbits
+; GCN-LABEL: bzhi32_d0:
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/extract-lowbits.ll:153]: This isn't canonical IR. We shouldn't be burning compile time optimizing non-canonical IR. Can you add a copy of this test with the instcombine output, and revert this if you get the same output

Comment (changpeng) [reply llvm/test/CodeGen/AMDGPU/extract-lowbits.ll:153]: Instcombine does not change the test source, and I think extracting the low bits are frequently used in the programs to restrict the range of a value.

File: llvm/lib/Target/AMDGPU/SIInstructions.td [hunk 3]
```diff
@@ -3553,6 +3553,23 @@ def : AMDGPUPat <
   (V_BFE_U32_e64 $src, (i32 0), $width)
 >;
 
+def uint5Bits : PatLeaf<(i32 VGPR_32:$width), [{
+  return CurDAG->computeKnownBits(SDValue(N, 0)).countMaxTrailingOnes() <= 5;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstructions.td:3557]: No, this is completely wrong!!! E.g. if bit 0 is known to be 0 then `countMaxTrailingOnes` will return 0 and this test will pass.

Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstructions.td:3557]:
```suggestion
  return CurDAG->computeKnownBits(SDValue(N, 0)).countMaxActiveBits() <= 5;
```
Please also add a test that shows why countMaxTrailingOnes was wrong.

Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIInstructions.td:3557]: Ugh, my bad, sorry.

Comment (changpeng) [reply llvm/lib/Target/AMDGPU/SIInstructions.td:3557]: Thanks so much.  Also countMaxActiveBits is more close to what we are going to define

Comment (changpeng) [reply llvm/lib/Target/AMDGPU/SIInstructions.td:3557]: Thanks, Done in https://github.com/llvm/llvm-project/pull/115543

File: llvm/test/CodeGen/AMDGPU/extract-lowbits.ll (hunk 2 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/extract-lowbits.ll:153]: Instcombine does change this: 


```
define i32 @bzhi32_d0(i32 %val, i32 %numlowbits) #0 {
  %numlow5bits = and i32 %numlowbits, 31
  %numhighbits = sub nuw nsw i32 32, %numlow5bits
  %1 = lshr i32 -1, %numhighbits
  %masked = and i32 %1, %val
  ret i32 %masked
}
```

Codegenning this does produce BFE already (except we fail to optimize out the clamp, which is a separate issue)

**Pull request merged**